### PR TITLE
[extension-selenium] Add the ability to use self-signed MITM for browser proxy

### PR DESCRIPTION
**Pull request by draker94**

### CODE REVIEW

File: vividus-extension-selenium/src/main/resources/properties/defaults/default.properties [hunk 1]
```diff
@@ -16,10 +16,12 @@ proxy.capture-types=REQUEST_CONTENT, RESPONSE_CONTENT, REQUEST_BINARY_CONTENT, R
 proxy.connectable-host=
 proxy.host=localhost
 proxy.mitm.enabled=false
+proxy.mitm.mitm-manager-type=IMPERSONATED
 proxy.mitm.key-store.alias=
 proxy.mitm.key-store.path=
 proxy.mitm.key-store.password=
 proxy.mitm.key-store.type=
+proxy.mitm.send-certificates=false
```
Comment (ikalinin1) [inline vividus-extension-selenium/src/main/resources/properties/defaults/default.properties:24]: please document this property

File: docs/modules/plugins/partials/proxy-properties.adoc [hunk 2]
```diff
@@ -4,6 +4,12 @@ a|`true`
 |`false`
 |Enable proxy for the whole test run
 
+|`proxy.recording.enabled`
+|`true`
```
Comment (valfirst) [inline docs/modules/plugins/partials/proxy-properties.adoc:8]:
```suggestion
a|`true`
```

File: docs/modules/plugins/partials/proxy-properties.adoc [hunk 3]
```diff
@@ -19,8 +25,57 @@ a|`true`
...
 a|`true`
 `false`
 |`false`
 |Enables publishing of attachments with HAR to each failed step. `@noHarOnFailure` meta tag can be used to disable HAR publishing at the story level.
+
+|`proxy.trust-all-servers`
+|`true`
+`false`
+|`false`
+|Whether to disable verification of upstream servers' SSL certificates
```
Comment (valfirst) [inline docs/modules/plugins/partials/proxy-properties.adoc:38]:
```suggestion
|Whether to disable verification of upstream servers SSL certificates
```

File: docs/modules/plugins/partials/proxy-properties.adoc [hunk 4]
```diff
@@ -19,8 +25,57 @@ a|`true`
...
+|`proxy.mitm.key-store.alias`
+|keystore alias
+|`<empty>`
+|The alias for certificate entry in keystore
+
+|`proxy.mitm.send-certificates`
+|`true`
+`false`
+|`false`
+|Whether to send certificates to servers (only for `SELF_SIGNED` MITM manager type)
```
Comment (ikalinin1) [inline docs/modules/plugins/partials/proxy-properties.adoc:81]: is there a real use-case for such a switch? Maybe you could use some default value for self-signed certs?

File: docs/modules/plugins/partials/proxy-properties.adoc [hunk 5]
```diff
@@ -19,8 +25,57 @@ a|`true`
...
+|`proxy.trust-all-servers`
+|`true`
+`false`
+|`false`
+|Whether to disable verification of upstream servers' SSL certificates
+
+|`proxy.capture-types`
+|HAR capture types
+|`REQUEST_CONTENT, RESPONSE_CONTENT, REQUEST_BINARY_CONTENT, RESPONSE_BINARY_CONTENT, REQUEST_HEADERS, RESPONSE_HEADERS`
+|List of HAR capture types
```
Comment (valfirst) [inline docs/modules/plugins/partials/proxy-properties.adoc:43]: it makes sense to list all available options here

File: docs/modules/plugins/partials/proxy-properties.adoc [hunk 6]
```diff
@@ -19,8 +25,57 @@ a|`true`
...
+
+|`proxy.mitm.enabled`
+|`true`
+`false`
+|`false`
+|Whether to enable https://docs.mitmproxy.org/stable[MITM proxy] for the whole test run
+
+|`proxy.mitm.mitm-manager-type`
+|`SELF_SIGNED`
```
Comment (valfirst) [inline docs/modules/plugins/partials/proxy-properties.adoc:52]:
```suggestion
a|`SELF_SIGNED`
```

File: docs/modules/plugins/partials/proxy-properties.adoc [hunk 7]
```diff
@@ -19,8 +25,57 @@ a|`true`
...
+`IMPERSONATED`
```
Comment (valfirst) [inline docs/modules/plugins/partials/proxy-properties.adoc:53]: it makes sense to describe the difference a bit

File: docs/modules/plugins/partials/proxy-properties.adoc [hunk 8]
```diff
@@ -19,8 +25,57 @@ a|`true`
...
+|`<empty>`
+|The password for keystore
+
+|`proxy.mitm.key-store.alias`
+|keystore alias
+|`<empty>`
+|The alias for certificate entry in keystore
+
+|`proxy.mitm.send-certificates`
+|`true`
```
Comment (valfirst) [inline docs/modules/plugins/partials/proxy-properties.adoc:78]:
```suggestion
a|`true`
```

File: docs/modules/plugins/partials/proxy-properties.adoc [hunk 9]
```diff
@@ -19,8 +25,57 @@ a|`true`
...
+|password for keystore
+|`<empty>`
+|The password for keystore
+
+|`proxy.mitm.key-store.alias`
+|keystore alias
+|`<empty>`
+|The alias for certificate entry in keystore
+
+|`proxy.mitm.send-certificates`
```
Comment (valfirst) [inline docs/modules/plugins/partials/proxy-properties.adoc:77]:
```suggestion
|`proxy.mitm.self-signed.send-certificates`
```

File: vividus-extension-selenium/src/main/java/org/vividus/proxy/mitm/MitmManagerOptions.java [hunk 10]
```diff
@@ -41,8 +51,19 @@ public boolean isTrustAllServers()
...
+    {
+        return sendCertificates;
+    }
+
     public KeyStoreOptions getKeyStoreOptions()
     {
         return keyStoreOptions;
     }
+
+    public enum MitmManagerType
```
Comment (valfirst) [inline vividus-extension-selenium/src/main/java/org/vividus/proxy/mitm/MitmManagerOptions.java:64]: move to a separate file

File: vividus-extension-selenium/src/main/java/org/vividus/proxy/mitm/MitmManagerFactory.java [hunk 11]
```diff
@@ -37,6 +41,21 @@ public MitmManager createMitmManager(MitmManagerOptions options)
...
 
+        MitmManager mitmManager;
+        if (options.getMitmManagerType() == MitmManagerType.IMPERSONATED)
+        {
+            mitmManager = buildImpersonatingMitmManager(options, keyStoreOptions);
+        }
+        else
+        {
+            mitmManager = buildSelfSignedMitmManager(options, keyStoreOptions);
+        }
```
Comment (valfirst) [inline vividus-extension-selenium/src/main/java/org/vividus/proxy/mitm/MitmManagerFactory.java:52]: move this logic to MitmManagerType, for example:

`MitmManagerType.IMPERSONATED.buildMitmManager(MitmManagerOptions options, KeyStoreOptions keyStoreOptions)`

**Review by valfirst — CHANGES_REQUESTED** [review]

File: docs/modules/plugins/partials/proxy-properties.adoc (hunk 4 above)
Comment (draker94) [reply docs/modules/plugins/partials/proxy-properties.adoc]: > is there a real use-case for such a switch? Maybe you could use some default value for self-signed certs?

Yes, indeed, I can hardly imagine a case where user will have to disable the sending of certificates. Using "true" as the default value should be enough

**Review by valfirst — APPROVED** [review]

**Review by ikalinin1 — APPROVED** [review]

**Review by uarlouski — APPROVED** [review]

**Pull request merged**